### ISSUE TITLE
[FEATURE] Image format support

### ISSUE THREAD

**Issue by aaronleopold** [stumpapp/stump#339]
**Describe the solution you'd like**

Add support for the following formats:

- [ ] AVIF
    - https://github.com/stumpapp/stump/pull/385 
- [ ] JPEG XL

**Additional context**

This mostly just involves implementing the `ImageProcessor` trait for each new format. I imagine the following libraries will be pulled in:

- https://github.com/tirr-c/jxl-oxide
- https://github.com/kornelski/cavif-rs

Please comment for any additional formats which should be supported if you feel I missed one

**Comment by smldub** [comment]
The [image crate](https://docs.rs/image/latest/image/index.html) implements AVIF natively, so I have added AVIF support (it works for with my test zipped AVIF files at least) by simply adding to the image/generic.rs file
```rust
...
let format = match options.format {
    process::ImageFormat::Jpeg => Ok(ImageFormat::Jpeg),
    process::ImageFormat::Png => Ok(ImageFormat::Png),
    process::ImageFormat::Avif => Ok(ImageFormat::Avif),
...
```
along with adding the AVIF objects throughout the image folder. But no need for a specific ImageProcessor. Let me know if you would like me to branch and pull this change (along with tests and stuff). Big rust newbie, so sorry if I am missing something that makes this a bad idea.

**Comment by aaronleopold** [comment]
Awesome! I didn't realize they added it in natively.

> Let me know if you would like me to branch and pull this change (along with tests and stuff). Big rust newbie, so sorry if I am missing something that makes this a bad idea.

Yeah, I'd appreciate any time you have to put up a PR. Be sure to work off of the `develop` branch. As for tests, there were recently a few added to that file which you can reference. Ideally, we want to assert that we can:

1. Convert other supported formats _to_ AVIF
2. Convert _from_ AVIF to other supported formats

No worries if it is missing something, I can help you get it to the finish line 🙂 

Also, I see a footnote in the [image repo](https://github.com/image-rs/image?tab=readme-ov-file#supported-image-formats) which might require a little investigation to see if the docker base images need any additional libs added or if they are baked in already:

> \* Requires the avif-native feature, uses the libdav1d C library.

**Comment by aaronleopold** [comment]
@smldub FYI I had to revert those changes 😞 I think the path forward is to find a solution which doesn't require an external library to be hooked up, or at the very least statically link the library so people don't have a hard requirement to install it.

I appreciate your time and efforts nonetheless! If you still want to try and integrate avif support with the above in mind, I'm here to help

**Comment by smldub** [comment]
Sorry for breaking everything! Maybe a possibility is to build with [conditional compilation](https://doc.rust-lang.org/reference/conditional-compilation.html)?

**Comment by aaronleopold** [comment]
> Sorry for breaking everything!

No it's all good! It wasn't just you, but it's also part of the process and the very reason we have `nightly` and `experimental` builds in the first place 🙂 

> Maybe a possibility is to build with [conditional compilation](https://doc.rust-lang.org/reference/conditional-compilation.html) (along with statically linking)

I think the problem with this route is that it would create a situation where we'd have to release two separate binaries: one with an (e.g.) `avif` feature flag enabled and one without it. In an ideal world, I think we would be able to do something more like what is [done for the PDFium library](https://github.com/stumpapp/stump/blob/main/core/src/filesystem/media/pdf.rs#L142-L160)

I'd sacrifice being able to generate AVIF thumbnails if there was a pure-rust decoder, but IIRC there only seems to be an encoder

**Comment by smldub** [comment]
>AVIF thumbnails if there was a pure-rust decoder

Sorry this is probably a stupid question, but if the only goal is to serve images to a web-browser, is it necessary to decode the image? When I first messed around with the image crate, just adding the avif format so that it could be loaded (without decoding) allowed me to see my files in the webapp.

**Comment by aaronleopold** [comment]
>  if the only goal is to serve images to a web-browser, is it necessary to decode the image?

No, not for the file serving aspect. That's just reading the entire file and dumping it to the client. 

You ask a good question though. Besides thumbnail generation, the only other functionality which I assumed would need a decoder would be grabbing the dimensions of the image during analysis. I don't know where in the header AVIF stores that information but we don't necessarily need a full decoder if we know the structure

**Comment by JMicheli** [comment]
> You ask a good question though. Besides thumbnail generation, the only other functionality which I assumed would need a decoder would be grabbing the dimensions of the image during analysis. I don't know where in the header AVIF stores that information but we don't necessarily need a full decoder if we know the structure
> 
> Edit to add a reference to the code where this happens: https://github.com/stumpapp/stump/blob/main/core/src/filesystem/media/analyze_media_job/task_analyze_dimensions.rs#L55-L61

This is something I've thought about. Image analysis is pretty slow right now because it loads the whole page to get its dimensions. I think that could be avoided in the future if we just read the dimensions info from the image files. We should determine 1) if image can do this or if another crate (or custom solution) is needed, and 2) if there are any additional difficulties resulting from many images being compressed into archive-based comic formats.

**Comment by aaronleopold** [comment]
> This is something I've thought about. Image analysis is pretty slow right now because it loads the whole page to get its dimensions. I think that could be avoided in the future if we just read the dimensions info from the image files. 

Yeah it's definitely a large use of resources just to grab X bytes from the header

> We should determine 1) if image can do this or if another crate (or custom solution) is needed

https://crates.io/crates/imagesize 🥹

> 2) if there are any additional difficulties resulting from many images being compressed into archive-based comic formats.

I wouldn't expect there to be, we can read precise amounts of bytes for any (supported) archive file entry, unless you mean something else?

**Comment by aaronleopold** [comment]
> Might be an easy fix, so I'll see if I can implement it in the near term

Yeah if image doesn't have anything for us that library I linked seems like it would just be a drop in replacement using the `blob_size` function, along with some logic to read the first 16ish bytes.

> > if the only goal is to serve images to a web-browser, is it necessary to decode the image?
> 
> No, not for the file serving aspect. That's just reading the entire file and dumping it to the client.

So @smldub you're welcome to try and re-add the support for just serving AVIF files to the client if you'd like, but no worries if not! I'll try to get to it this weekend so you aren't stuck with AVIF images not being served

**Comment by smldub** [comment]
Some thoughts on Jpeg_XL, I don't think the format will be added to image [for quite some time](https://github.com/image-rs/image/issues/1979), so some options for a processor are
1. Use jxl-oxide for decoding, then plug that into image-rs as a bitmap for rescaling and reformatting. The images could be re-encoded using [zune-jpegxl](https://github.com/etemesi254/zune-image/tree/dev/crates/zune-jpegxl) or you could just make thumbnails into a different format. 
2. Use [zune-image](https://github.com/etemesi254/zune-image) for both decoding, encoding, and scaling (uses jxl-oxide as a decoder). Zune-jpeg is currently how image-rs handles [jpeg decoding](https://github.com/image-rs/image/blob/main/src/codecs/jpeg/decoder.rs), so there is code on how to convert from zune images into image-rs images, but that might not even be necessary as it supports many of the same formats ([Supported zune-image formats](https://github.com/etemesi254/zune-image/tree/dev/crates/zune-image)).

Let me know if you have any thoughts (I am leaning towards 1), I will probably mess around with this issue for the next little bit!

**Comment by aaronleopold** [comment]
I'll try to give that linked issue a read over the weekend, looks like a lot of good context in there

Wrt the two options, I think it mostly depends on whether jxl will be a support format thumbnail generation. Assuming we can get the imagesize crate to function as expected, if a format isn't going to be supported for thumbnail generation we really don't need an encoder AFAIK. However, I did notice that all of the formats supported by `zune-image` are feature flagged, so we could just pull in the jxl (or whatever it might be named) feature to have encoding/decoding in one place. 

I don't think I really have a preference, whatever is the path of least resistance is likely what I'd say to go for. This format doesn't require any external libs, so we should be good either way 😅

**Comment by Quackdoc** [comment]
Was just shown this project, looks neat and I for sure plan on trying it soon! 

AVIF support in image-rs is limited to YUV 420,422,444 decode,  as for encoding it uses rav1e which is quite slow, but at least it's native rust, these two issues might be the cause of some of the issues you had run into. Rav1d is a rust native decoder which is very much a WIP, but will allow for native rust decoding without the need for any external libraries, and should hopefully be wired up not too far into the future. Will not solve the yuv issues. Grayscale (YUV400) 10bit+ wont be magically fixed by this. 

as for JXL, if you are doing encoding within rust, take note that zune-jxl's encoder is fairly basic, I'm not sure using it for encoding a thumbnail is going to make a lot of sense. zune-image works great with jxl decoding, and decoding in general, I found it marginally faster then image-rs, using jxl-oxide directly is still the fastest I find however I haven't checked for a bit. I personally store all my comics in jxl so it would be nice to have support for that.

**Comment by aaronleopold** [comment]
> Was just shown this project, looks neat and I for sure plan on trying it soon!

Nice! I hope you wind up enjoying it. Do be weary of some of the recent bugs and some rough edges, and feel free to report any issues you run into.

> AVIF support in image-rs is limited to YUV 420,422,444 decode, as for encoding it uses rav1e which is quite slow, but at least it's native rust, these two issues might be the cause of some of the issues you had run into. Rav1d is a rust native decoder which is very much a WIP, but will allow for native rust decoding without the need for any external libraries, and should hopefully be wired up not too far into the future. Will not solve the yuv issues. Grayscale (YUV400) 10bit+ wont be magically fixed by this.

I really appreciate the context here. I think the yuv issues are acceptable on paper, but it would be good to know how images converted from AVIF to any in the current supported format set for thumbnail generation would actually turn out. Otherwise, the only other use case for decoding would be during specific analysis tasks where the color space is not important (e.g. image dimensions).

> as for JXL, if you are doing encoding within rust, take note that zune-jxl's encoder is fairly basic, I'm not sure using it for encoding a thumbnail is going to make a lot of sense. zune-image works great with jxl decoding, and decoding in general, I found it marginally faster then image-rs, using jxl-oxide directly is still the fastest I find however I haven't checked for a bit. 

Again, I really appreciate the context here! I think much of the same applies here as with AVIF. I guess another thing to note is the linked crate above, `imagesize`, could be used to get the image dimensions without needing to decode the image. Still doesn't solve the issue of if you have a JXL/AVIF library and want to opt into generating thumbnails 😅 
 
> I personally store all my comics in jxl so it would be nice to have support for that.

Support for JXL file serving at the very least was merged just today, coincidentally. It would be in the latest `nightly` build, and will go out in `0.0.6` along with a few other bug fixes

**Comment by aaronleopold** [comment]
I'm going to close this for now since I haven't heard much demand for AVIF support in a while. If anyone would like to contribute support for it, though, I would very much appreciate it!
